llarp/dnsd.cpp

```diff
@@ -432,38 +438,58 @@ raw_handle_recvfrom(int *sockfd, const struct sockaddr *saddr, const void *buf,
 }
 
 bool
-llarp_dnsd_init(struct dnsd_context *dnsd, struct llarp_logic *logic,
-                struct llarp_ev_loop *netloop, const char *dnsd_ifname,
-                uint16_t dnsd_port, const char *dnsc_hostname,
-                uint16_t dnsc_port)
+llarp_dnsd_init(struct dnsd_context *const dnsd,
+                struct llarp_logic *const logic,
+                struct llarp_ev_loop *const netloop,
+                const llarp::Addr &dnsd_sockaddr,
+                const llarp::Addr &dnsc_sockaddr)
 {
-  struct sockaddr_in bindaddr;
-  bindaddr.sin_addr.s_addr = inet_addr("0.0.0.0");
+  // struct sockaddr_in bindaddr;
+  /*
+  llarp::Addr dnsd_addr;
+  bool ifRes = GetIFAddr(std::string(dnsd_ifname), dnsd_addr, AF_INET);
+  if (!ifRes)
+  {
+    llarp::LogError("Couldn't init dns server, can't resolve interface: ",
+  dnsd_ifname); return false;
+  }
+  llarp::LogInfo("DNSd interface ip ", dnsd_addr);
+  */
+
+  /*
+  bindaddr.sin_addr.s_addr = inet_addr("127.0.0.1"); // network byte
   bindaddr.sin_family      = AF_INET;
   bindaddr.sin_port        = htons(dnsd_port);
+  */
 
   dnsd->udp.user     = &dns_udp_tracker;
   dnsd->udp.recvfrom = &llarp_handle_dns_recvfrom;
   dnsd->udp.tick     = nullptr;
 
   dns_udp_tracker.dnsd = dnsd;
 
-  dnsd->tracker = &dns_udp_tracker;  // register global tracker with context
-  // dnsd->logic     = logic;             // set logic worker for timers
-  dnsd->intercept = nullptr;  // set default intercepter
+  dnsd->tracker   = &dns_udp_tracker;  // register global tracker with context
+  dnsd->intercept = nullptr;           // set default intercepter
+
+  // set up fresh socket
+  dnsd->client.udp           = new llarp_udp_io;
```

Comment:
`llarp_udp_io` doesn't need to be a pointer on a client, it can and probably should be a member as not a pointer. you can always do `&dnsd->client.udp` instead and it'll work. better that way so you don't have to remember to `delete` it

Reply:
I understand that but I would still chose a pointer here. The client may needs to reuse the daemon socket (as the previous design needed), or multiple clients may need to run on the same socket. I'm favoring the flexibility here, as I envision it as more of a library than something hard coded to our specific use (trying to keep the lokinet specifics to inside the hooks). And we're talking leaks no more than 10 pointers here that likely only need to be cleaned up on exit but I understand your general principle.